### PR TITLE
safrole test

### PR DESCRIPTION
**Pull request by turuslan**
- Names (`mathcal_O`, `banderout`, ...) are taken from graypaper latex

### CODE REVIEW

File: test-vectors/safrole/safrole.hpp [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+  }
+
+  // [JAM:273]
+  template <size_t N>
+  auto de(qtils::BytesN<N> x) {
+    return boost::endian::
+        endian_load<uint64_t, N, boost::endian::order::little>(x.data());
+  }
+
+  // [JAM:3.8.1]
```
Comment (kamilsa) [inline test-vectors/safrole/safrole.hpp:89]: These references to graypaper can quickly become obsolete. I think it makes sense to write GP version next to each reference

File: test-vectors/safrole/safrole.hpp [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+    if (E_T.size() > K) {
+      throw std::logic_error("not covered by test vectors");
+    }
+    if (m_tick >= Y and not E_T.empty()) {
+      return std::make_pair(
+          state, typename types::Output{Error::unexpected_ticket});
+    }
+
+    // [JAM:59]
+    // TODO: offenders
```
Comment (kamilsa) [inline test-vectors/safrole/safrole.hpp:224]: Please create an issue in this repo

File: test-vectors/safrole/safrole.hpp [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+    }
+    return out;
+  }
+
+  struct Epoch {
+    uint32_t epoch, phase;
+    Epoch(uint32_t slot) : epoch{slot / E}, phase{slot % E} {}
+  };
+
+  static std::pair<typename types::State, typename types::Output> transition(
```
Comment (kamilsa) [inline test-vectors/safrole/safrole.hpp:189]: I think it is worth to leaving a comment for this function

File: test-vectors/safrole/safrole.test.cpp [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+  DIFF_M(gamma_a);
+  DIFF_M(gamma_s);
+  DIFF_M(gamma_z);
+}
+
+/**
+ * Check safrole state transition against test vectors.
+ */
+template <bool full>
+void test_transition() {
```
Comment (kamilsa) [inline test-vectors/safrole/safrole.test.cpp:210]: Can't we use Gtest here?

**Review by kamilsa — APPROVED** [review]

**Review by kamilsa — COMMENTED** [review]
Check comments above

**Review by xDimon — APPROVED** [review]

**Pull request merged**